### ISSUE TITLE
[discussion] Include Environment Variables

### ISSUE THREAD

**Issue by ianlewis** [slsa-framework/slsa-github-generator#822]
# Background:

For reproducibility, build steps are included in provenance by some builders. These build steps include information to reproduce the command, such as the working directory and environment variables.

POSIX [defines some environment variables](https://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html) which mirror values set elsewhere. These include the `PWD`, `USER`, and `HOME` environment variables. Some programs may use the environment variables as a way to determine the current working director or current user for example.

Some values, such as the current working directory, are set in a separate field from the environment variables and thus the information in the environment variables would be redundant.

# Question

Should these environment variables be included in the build steps in the provenance? If so, which ones?

**Comment by ianlewis** [comment]
Currently we are setting the `PWD` environment variable but are leaning towards removing it since it's redundant info and we can say that our `buildConfig` implicitly indicates that it should be set when reproducing the build step.

**Comment by laurentsimon** [comment]
@TomHennen @MarkLodato @joshuagl @msuozzo

**Comment by mlieberman85** [comment]
+1 to removing it.  Does https://github.com/slsa-framework/slsa-github-generator/generic@v1 have a defined spec yet? Just making that clear in the spec would be useful.

**Comment by laurentsimon** [comment]
We don't have a "spec", just examples https://github.com/slsa-framework/slsa-github-generator/blob/main/internal/builders/generic/README.md#provenance-format

For go builder we have *some* explanation in https://github.com/slsa-framework/slsa-github-generator/tree/main/internal/builders/go#buildconfig-format, but it's probably not enough to explain what is recorded, what is not, and maybe take the view of a re-builder?

Great idea to have a spec, thanks! All the builders will have the same output, so it should be easy. 

What would you like to see captured in the specs?

**Comment by shaunmlowry** [comment]
I think it's dangerous to start arbitrarily removing environment variables from provenance just because they're redundant on a specific platform (or set of platforms).  What about build steps that run on Windows where PWD isn't a thing, and might validly be used for some user-defined purpose by such a build step?

**Comment by laurentsimon** [comment]
Note that we have not released our builders with `PWD` set, this is just a change we made a few days ago, but it's not released, which is why we're removing it.

I think you also need more context: we don't record env variables, we only record the ones the user asked us to set, and we don't record those that exist "by default". The user never asked us to set `PWD` in this case.

Recording all env variables is something we could do. In general we would not want to capture secrets, but in our case we don't have any.

**Comment by shaunmlowry** [comment]
I think _anything_ that could potentially affect the output of the build in any way needs to be recorded, e.g. PATH, LANG, LC_* could all have subtle effects on the output artifact

**Comment by ianlewis** [comment]
I tend to agree with @shaunmlowry. I think we should set `PWD`. The question is really whether we need to record it in the provenance. Though we do need to consider how much noise is added to the provenance, if it's set I think it should probably be included to make it more explicit.

Obviously we separately endeavor to not include secrets in the recorded environment variables but these variables are those set when the command is executed (any set by the build command itself aren't visible to provenance generation. You'd have to include secrets in your environment set in your goreleaser config or something like that for it to be potentially included here. However, we could sanitize obvious ones like `AWS_SECRET_ACCESS_KEY` etc. so it's obvious they were set, but their values are obscured. It's a bit hard to know what's a secret more generally though.

**Comment by mlieberman85** [comment]
@shaunmlowry brings up a good point. I'm ok with some redundancy, especially in cases where someone might be doing something strange to normally read-only env vars.

@ianlewis I'm not super familiar with how the reusable workflows work, but could someone configure it with a list of environment variables to not include in the output? Give it to the user what should/shouldn't be included in provenance.

**Comment by ianlewis** [comment]
I agree that, while I argued for it's inclusion, `PWD` is probably not really important to include and that `PWD` would probably just be noise. Most likely any others that have an analogous field in the `buildConfig` making it redundant would just be noise too. I'm not sure that I would feel comfortable saying that for every variable we might set though. We might need to consider them on a case-by-case basis and document them in some way. In practice though, I hope that few users will really need to care.

> @ianlewis I'm not super familiar with how the reusable workflows work, but could someone configure it with a list of environment variables to not include in the output? Give it to the user what should/shouldn't be included in provenance.

This is possible. Though the inputs to a workflow aren't structured (only strings, integers, etc.) maybe it could be supported via a config file that is included in the repo. I'm not sure we're at the point of considering this option yet though.
